Lib/Src/CUartCom.cpp

```diff
@@ -10,96 +10,231 @@
+    // Check event (empty message or not)
+    if (msg.empty())
+    {
+        event = EMPTY_MSG;
+    }
+    // Push msg to queue
+    if (m_tx_queue.size() <= MAX_TX_QUEUE_SIZE && event == NON_EMPTY_MSG)
     {
-        HAL_GPIO_WritePin(m_uart_de_port, m_uart_de_pin, GPIO_PIN_SET);
+        m_tx_queue.push(msg);
     }
 
-    // Convert string to C style
-    const char *c_msg = msg.c_str();
+    // Check state
+    if (mb_state == IDLE)
+    {
+        if (event == EMPTY_MSG)
+        {
+            // Disable interrupt, disable pin
+            __HAL_UART_DISABLE_IT(mp_huart, UART_IT_TXE);
+            m_uart_de_pin.set(false);
+        }
+        else
+        {
+            // Enable USART_DE pin
+            m_uart_de_pin.set(true);
+
+            // Start transmission with interrupts
+            std::string message = m_tx_queue.front();
+            uint8_t msg_length = message.length();
+            strcpy(m_tx_buffer, message.c_str());
+            m_tx_queue.pop();
+            HAL_UART_Transmit_IT(mp_huart, (uint8_t *)m_tx_buffer, msg_length);
+
+            // Change state to TX
+            mb_state = TX;
+        }
+    }
+    else
+    {
+        if (event == EMPTY_MSG)
+        {
+            // Transmit next element in queue
+            std::string message = m_tx_queue.front();
+            uint8_t msg_length = message.length();
+            strcpy(m_tx_buffer, message.c_str());
+            m_tx_queue.pop();
+            HAL_UART_Transmit_IT(mp_huart, (uint8_t *)m_tx_buffer, msg_length);
+        }
+    }
+}
 
-    // Check it is a valid string and send error message if not
-    if (!(*c_msg) || c_msg == nullptr)
+/**
+ * @brief UART Reception Complete Interrupt Handler
+ * @arg p_huart Pointer to UART Handler
+ */
+void CUartCom::uartRxHandler(UART_HandleTypeDef *p_huart)
+{
+    // Store incoming char
+    if (m_rx_buffer.put((const char)m_rx_char))
     {
-        const uint8_t error_msg_len = 28;
-        const char *error_msg = "Invalid string from c_str()\n";
-        HAL_UART_Transmit(mp_huart,
-                          (uint8_t *)error_msg,
-                          error_msg_len,
-                          UART_TIMEOUT);
-
-        // Disable UART_DE Pin
-        if (m_uart_de_port != nullptr)
+        // If end of string or rx_buffer is full, add to queue
+        std::string command_string = m_rx_buffer.get();
+
+        // If the queue has reached its max size, the message will be
+        // dismissed
+        if (m_rx_queue.size() <= MAX_RX_QUEUE_SIZE && !command_string.empty())
         {
-            HAL_GPIO_WritePin(m_uart_de_port, m_uart_de_pin, GPIO_PIN_RESET);
+            m_rx_queue.push(command_string);
         }
+    }
+
+    HAL_UART_Receive_IT(p_huart, &m_rx_char, 1);
+}
+
+/**
+ * @brief UART Transmission Complete Interrupt Handler
+ * @arg p_huart Pointer to UART Handler
+ */
+void CUartCom::uartTxHandler(UART_HandleTypeDef *p_huart)
+{
+    // Check if queue is empty
```

Comment:
You can pop() the queue here. For example: `m_tx_buffer` can be turned into a pointer and if it's nullptr, then no need to pop(). If it is not empty, then pop the m_tx_queue and then set m_tx_buffer to nullptr.

Reply:
This relates to my previous comment about how you organised the retrieval of data from `m_tx_queue`. You `strcpy()` the data out. But you could just use the content of the queue directly and then `pop()` the queue when transmission is complete.
Perhaps, another pair programming session is in order. Let me know if you want it.

Reply:
I can't use the data directly because the queue stores strings and I need to send c-type strings. That's why I'm copying the data to the buffer. Another pair programming session would be great, the last one was very helpful.